### PR TITLE
General refactor & polish

### PR DESCRIPTION
**Pull request by furiousdom**

### CODE REVIEW

File: routes/users.js [hunk 1]
```diff
@@ -5,6 +5,12 @@ const passport = require('passport');
 
 const User = require('../models/User');
 
+const msg = {
+  fillFields: 'Please fill all fields.',
+  failedMatch: 'Passwords don\'t match.',
+  short: 'Password should be at least 8 characters.'
+};
+
```
Comment (lukasaric) [inline routes/users.js]: - Create a controller file for users (you can name it `users.controller.js`).
- Then extract all router callbacks (login, register...) there, and import it here.
- Then we will polish all callback functions after :bowtie:

File: config/passport.js [hunk 2]
```diff
@@ -3,36 +3,35 @@ const bcrypt = require('bcryptjs');
...
-            }
-          });
-        })
-        .catch(err => console.log(err));
+const msg = {
+  wrongPassword: 'Password incorrect.',
+  notRegistered: 'That email is not registered.'
+};
+
+const authenticateUser = (email, password, done) => {
```
Comment (lukasaric) [inline config/passport.js]: You can name it `verifyLocal`. Also, this can be a regular function.

Comment (furiousdom) [reply config/passport.js]: When you say regular function, you mean not an arrow function, correct?

Comment (lukasaric) [reply config/passport.js]: Yes. In this case, there's no need for arrow function. There is no kind of `this` manipulation, and also function is not short enough to inline it (like in serialization).

File: config/passport.js [hunk 3]
```diff
@@ -8,26 +7,20 @@ const msg = {
   notRegistered: 'That email is not registered.'
 };
 
-const authenticateUser = (email, password, done) => {
-  return User.findOne({ email: email })
+function verifyLocal(email, password, done) {
+  return User.findOne({ email })
     .then(user => {
       if (!user) return done(null, false, msg.notRegistered);
```
Comment (lukasaric) [inline config/passport.js]: Don't leave an unnecessary empty line. Move line bellow here.

File: config/passport.js [hunk 4]
```diff
@@ -8,26 +7,20 @@ const msg = {
...
-      bcrypt.compare(password, user.password, (err, isMatch) => {
-        if (err) throw err;
-
-        if (isMatch) {
-          return done(null, user);
-        } else {
-          return done(null, false, msg.wrongPassword);
-        }
-      });
+      return user.authenticate(password)
```
Comment (lukasaric) [inline config/passport.js]: hmm, if you want to inline this but keep ruler border rule you can do something like this:

```js
const errorArgs = [null, false, msg.wrongPassword];
return user.authenticate(password) ? done(null, user) : done(...errorArgs);
```

Still, it's overkill, but for educational purpose good 😉

File: controllers/users.controller.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    text: 'You are now registered and can log in.'
+  },
+  logoutComplete: {
+    label: 'success_msg',
+    text: 'You are logged out.'
+  }
+};
+
+function register(req, res) {
+  const { name, email, password, password2 } = req.body;
```
Comment (lukasaric) [inline controllers/users.controller.js]: - You destructure it directly when receiving arguments.
- password2? 🤡 Rename it, something like: `rePassword`, `confirmPassword`

File: controllers/users.controller.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  let errors = [];
+
+  if (!name || !email || !password || !password2) errors.push(msg.fillFields);
+
+  if (password !== password2) errors.push(msg.failedMatch);
+
+  if (password.length < 8) errors.push(msg.short);
+
+  if (errors.length > 0) {
+    res.render('register', {
```
Comment (lukasaric) [inline controllers/users.controller.js]: - Above `if` statement, create object named `payload` and assign all data to it
- Then put it as the second parameter in `render` method
- Then inline `if` statement

File: controllers/users.controller.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+          });
+        } else {
+          const newUser = new User({
+            name,
+            email,
+            password
+          });
+
+          // eslint-disable-next-line handle-callback-err
+          bcrypt.genSalt(10, (err, salt) =>
```
Comment (lukasaric) [inline controllers/users.controller.js]: Extract it as an instance method (`encrypt`).

File: routes/users.js [hunk 8]
```diff
@@ -1,86 +1,13 @@
...
-const msg = {
-  fillFields: 'Please fill all fields.',
-  failedMatch: 'Passwords don\'t match.',
-  short: 'Password should be at least 8 characters.'
-};
+const { register, login, logout } = require('../controllers/users.controller');
 
 router.get('/login', (req, res) => res.render('login'));
 router.get('/register', (req, res) => res.render('register'));
+router.get('/logout', (req, res) => logout(req, res));
```
Comment (lukasaric) [inline routes/users.js]: Just leave it as: `router.get('/logout', logout);`
JS does the rest of magic 😉

File: routes/users.js [hunk 9]
```diff
@@ -1,86 +1,13 @@
...
-  })(req, res, next);
-});
-
-router.get('/logout', (req, res) => {
-  req.logout();
-  req.flash('success_msg', 'You are logged out.');
-  res.redirect('/users/login');
-});
+router.post('/register', (req, res) => register(req, res));
+router.post('/login', (req, res, next) => login(req, res, next));
```
Comment (lukasaric) [inline routes/users.js]: See the previous comment 🔝

File: controllers/users.controller.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      password,
+      password2
+    });
+  } else {
+    return User.findOne({ email })
+      .then(user => {
+        if (user) {
+          errors.push(msg.exists);
+          res.render('register', {
+            errors,
```
Comment (lukasaric) [inline controllers/users.controller.js]: Reuse the payload object here.

File: config/passport.js (hunk 4 above)
Comment (furiousdom) [reply config/passport.js]: In my opinion it is better the way it is now, so I will leave it the way it is.

File: controllers/users.controller.js (hunk 5 above)
Comment (furiousdom) [reply controllers/users.controller.js]: For the first point I have to say what about line 66, I use variable req there, the second point I handled.

Comment (lukasaric) [reply controllers/users.controller.js]: Good point, I missed that 🙃

File: controllers/users.controller.js [hunk 11]
```diff
@@ -19,22 +18,24 @@ const msg = {
 };
 
 function register(req, res) {
-  const { name, email, password, password2 } = req.body;
+  const { name, email, password, rePassword } = req.body;
   // eslint-disable-next-line prefer-const
   let errors = [];
```
Comment (lukasaric) [inline controllers/users.controller.js]: Extract all error handling logic as `errorHandler` function above.

File: controllers/users.controller.js [hunk 12]
```diff
@@ -51,18 +52,13 @@ function register(req, res) {
...
-              newUser.password = hash;
-              newUser.save()
-                .then(user => {
-                  req.flash(msg.regComplete.label, msg.regComplete.text);
-                  res.redirect('/users/login');
-                })
-                .catch(err => console.log(err));
-            }));
+          newUser.encrypt();
+          newUser.save()
```
Comment (lukasaric) [inline controllers/users.controller.js]: Bonus task:
After you apply requested changes, try to use `save` user changes directly in `encrypt` method.
Then encrypt should return new user (`.then(user...`) here.

File: controllers/users.controller.js [hunk 13]
```diff
@@ -0,0 +1,81 @@
+const passport = require('passport');
+
+const User = require('../models/User');
+
+const msg = {
```
Comment (lukasaric) [inline controllers/users.controller.js]: Decouple all messages in one single file (in `config` folder). Then import it here & where it's needed.

File: controllers/users.controller.js [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    label: 'success_msg',
+    text: 'You are now registered and can log in.'
+  },
+  logoutComplete: {
+    label: 'success_msg',
+    text: 'You are logged out.'
+  }
+};
+
+function errorHandler({ name, email, password, rePassword }) {
```
Comment (lukasaric) [inline controllers/users.controller.js]: - Extract `errorHandler` as middleware function in `users` router.
- Hint: `router.post('/register', errorHandler, register);`
In function you will assign `errors` to `req.errors = errors`.
Then call `next()`.

P.S remove all empty lines in this function.

File: controllers/users.controller.js [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+  return {
+    errors,
+    name,
+    email,
+    password,
+    rePassword
+  };
+}
+
+function register(req, res) {
```
Comment (lukasaric) [inline controllers/users.controller.js]: - After you apply ☝️, in `req` you will have `errors` and `body`
- Then assign to payload what all credentials & errors

File: controllers/users.controller.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  const payload = errorHandler(req.body);
+
+  if (payload.errors.length > 0) res.render('register', payload);
```
Comment (lukasaric) [inline controllers/users.controller.js]:
```js 
if (payload.errors.length) return res.render('register', payload);
```
- Then remove `else` one line bellow

File: controllers/users.controller.js [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+  else {
+    const { name, email, password } = payload;
+    return User.findOne({ email })
+      .then(user => {
+        if (user) {
+          payload.errors.push(msg.exists);
+          res.render('register', payload);
```
Comment (lukasaric) [inline controllers/users.controller.js]: - `return` it
- Then remove `else`

File: controllers/users.controller.js [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        } else {
+          const newUser = new User({
```
Comment (lukasaric) [inline controllers/users.controller.js]:
```js
const newUser = new User({ name, email, password });
```

File: controllers/users.controller.js [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+            name,
+            email,
+            password
+          });
+
+          newUser.encryptPassword()
```
Comment (lukasaric) [inline controllers/users.controller.js]: `return` promise

File: controllers/users.controller.js [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+            .then(user => {
```
Comment (lukasaric) [inline controllers/users.controller.js]: `user` is unused.
- `.then(() => ...` or `.then(_user => ...`

File: models/User.js [hunk 21]
```diff
@@ -19,6 +22,18 @@ const UserSchema = new mongoose.Schema({
...
 });
 
+UserSchema.methods.authenticate = function (password) {
+  return bcrypt.compare(password, this.password)
+    .then(isMatch => isMatch ? this : false);
+};
+
+const encrypt = password => bcrypt.hash(password, SALT);
+
+UserSchema.methods.encryptPassword = async function () {
```
Comment (lukasaric) [inline models/User.js]: Neat 👀 🚀

File: routes/users.js [hunk 22]
```diff
@@ -1,86 +1,13 @@
...
-  })(req, res, next);
-});
-
-router.get('/logout', (req, res) => {
-  req.logout();
-  req.flash('success_msg', 'You are logged out.');
-  res.redirect('/users/login');
-});
+router.post('/register', register);
+router.post('/login', login);
```
Comment (lukasaric) [inline routes/users.js]: See comment for using `errorHandler` as middleware.

File: config/passport.js [hunk 23]
```diff
@@ -1,38 +1,25 @@
...
+        : done(null, false, msg.wrongPassword);
     })
-  );
+    .catch(err => done(err));
+}
 
-  passport.serializeUser((user, done) => {
-    done(null, user.id);
-  });
+module.exports = (passport) => {
```
Comment (lukasaric) [inline config/passport.js]: Remove unnecessary brackets and empty lines bellow.

File: controllers/users.controller.js [hunk 24]
```diff
@@ -0,0 +1,40 @@
+const passport = require('passport');
+
```
Comment (lukasaric) [inline controllers/users.controller.js]: Remove empty line, and order `require`s alphabetically here and everywhere else.
- You can install this [eslint-require-sort](https://github.com/zcuric/eslint-plugin-require-sort) packet authored by my college. 😉

File: controllers/users.controller.js [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+const User = require('../models/User');
+const msg = require('../config/messages');
+
+function register(req, res) {
+  const { errors, body: { name, email, password, rePassword } } = req;
+  const payload = { errors, name, email, password, rePassword };
+  if (errors.length > 0) return res.render('register', payload);
```
Comment (lukasaric) [inline controllers/users.controller.js]: `if (errors.length) ...` is enough

File: controllers/users.controller.js [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+        payload.errors.push(msg.exists);
+        return res.render('register', payload);
+      }
+      const newUser = new User({ name, email, password });
+      return newUser.encryptPassword()
+        .then(() => {
+          req.flash(msg.regComplete.label, msg.regComplete.text);
+          res.redirect('/users/login');
+        })
+        .catch(err => console.log(err));
```
Comment (lukasaric) [inline controllers/users.controller.js]: `.catch(err => next(err));`

File: controllers/users.controller.js [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+    });
+}
+
+function login(req, res, next) {
```
Comment (lukasaric) [inline controllers/users.controller.js]: You can remove `login` cuz you don't do anything more than authenticating with passport.
- Return it to `user routes` and extract options on top:
```js
router.post('/login', passport.authenticate('local', authenticateOptions));
```

File: routes/users.js [hunk 28]
```diff
@@ -1,86 +1,24 @@
 const express = require('express');
+const msg = require('../config/messages');
 const router = express.Router();
-const bcrypt = require('bcryptjs');
-const passport = require('passport');
 
-const User = require('../models/User');
+const { register, login, logout } = require('../controllers/users.controller');
+
+function errorHandler(req, _, next) {
```
Comment (lukasaric) [inline routes/users.js]: Helper functions (like this one) should go bellow `module.export`

**Pull request merged**